apps/src/templates/progress/googlePlatformApiRedux.js

```diff
@@ -0,0 +1,139 @@
+import firehoseClient from '@cdo/apps/lib/util/firehose';
+
+/*
+ * The Google Classroom Share Button is only available through the google
+ * platform api, so we have to add it to our page on load. We manage loading
+ * in redux since multiple components need to observe the loading state.
+ */
+
+const START_LOADING_GAPI = 'googlePlatformApi/START_LOADING_GAPI';
+const FINISH_LOADING_GAPI = 'googlePlatformApi/FINISH_LOADING_GAPI';
+
+const startLoadingGapi = startTime => ({type: START_LOADING_GAPI, startTime});
+const finishLoadingGapi = success => ({type: FINISH_LOADING_GAPI, success});
+
+export const GOOGLE_PLATFORM_API_ID = 'GooglePlatformApiId';
+const LOAD_TIMEOUT_MILLIS = 15000;
+
+const initialState = {
+  loading: false,
+  loaded: false,
+  loadStartTime: null
+};
+
+export function loadGooglePlatformApi() {
+  return (dispatch, getState) => {
+    if (gapiReady()) {
+      dispatch(finishLoadingGapi(true));
+      return Promise.resolve();
+    } else {
+      dispatch(startLoadingGapi(Date.now()));
+      return new Promise((resolve, reject) => {
+        const promise = {resolve: resolve, reject: reject};
+        loadApi(dispatch, getState, promise);
+      });
+    }
+  };
+}
+
+function onLoadFinished(success, promise) {
+  return (dispatch, getState) => {
+    dispatch(finishLoadingGapi(success));
+    if (success) {
+      promise.resolve();
+    } else {
+      promise.reject('Google Platform API failed to load.');
+    }
+
+    const data = {
+      success: success,
+      load_time: elapsedLoadTimeSeconds(getState)
+    };
+    firehoseClient.putRecord(
+      {
+        study: 'google-classroom-share-button',
+        study_group: 'v0',
+        event: 'api_load_finished',
+        data_json: JSON.stringify(data)
+      },
+      {includeUserId: true}
+    );
+  };
+}
+
+export function canShowGoogleShareButton(state) {
+  return !!state.googlePlatformApi && state.googlePlatformApi.loaded;
+}
+
+export default function googlePlatformApi(state = initialState, action) {
+  if (action.type === START_LOADING_GAPI) {
+    return {
+      ...state,
+      loading: true,
+      loadStartTime: action.startTime
+    };
+  }
+  if (action.type === FINISH_LOADING_GAPI) {
```

Comment:
This one is a bit confusing to me. If I finished loading the api, but the api errored, should `loaded` be true or false? Here it seems we are overloading `loaded` to mean "successfully loaded"

Reply:
yes, that's what it means -- i.e. "the api is loaded". i initially had another state property for `failed`, but a) it was never used, and b) setting `loaded = action.success; failed = !action.success` on consecutive lines seemed redundant, so i removed it. but i can add it back if you think it would make this clearer. or i can rename it to `successfullyLoaded`, but i hoped that was implied by `loaded = action.success`. but since it made you pause, i should probably rename it?

if our redux store persisted between route changes i think there would be more value in keeping track of failed loads in case we want to stop trying or something like that. but alas this state is super local and transient, so i decided to err on the side of simplicity.

Reply:
I was originally going to suggest two action types: "LOADING_GAPI_SUCCES" and "LOADING_GAPI_FAIL" or something similar.

Reply:
But I'm downgrading this comment to a NIT :)

Reply:
haha, one of the guides i read specifically called out that whether to separate those into separate actions or include a bool value is a matter of preference. since you've downgraded the importance of this i'm going to leave it as is. :)